hbase-common/src/main/java/org/apache/hadoop/hbase/io/crypto/tls/HBaseTrustManager.java

```diff
@@ -0,0 +1,160 @@
+  }
+
+  @Override
+  public X509Certificate[] getAcceptedIssuers() {
+    return x509ExtendedTrustManager.getAcceptedIssuers();
+  }
+
+  @Override
+  public void checkClientTrusted(X509Certificate[] chain, String authType, Socket socket)
+    throws CertificateException {
+    x509ExtendedTrustManager.checkClientTrusted(chain, authType, socket);
+    if (clientHostnameVerificationEnabled) {
+      performHostVerification(socket.getInetAddress(), chain[0]);
+    }
+  }
+
+  @Override
+  public void checkServerTrusted(X509Certificate[] chain, String authType, Socket socket)
+    throws CertificateException {
+    x509ExtendedTrustManager.checkServerTrusted(chain, authType, socket);
+    if (serverHostnameVerificationEnabled) {
+      performHostVerification(socket.getInetAddress(), chain[0]);
+    }
+  }
+
+  @Override
+  public void checkClientTrusted(X509Certificate[] chain, String authType, SSLEngine engine)
+    throws CertificateException {
+    x509ExtendedTrustManager.checkClientTrusted(chain, authType, engine);
+    if (clientHostnameVerificationEnabled) {
+      try {
+        performHostVerification(InetAddress.getByName(engine.getPeerHost()), chain[0]);
+      } catch (UnknownHostException e) {
+        throw new CertificateException("Failed to verify host", e);
+      }
+    }
+  }
+
+  @Override
+  public void checkServerTrusted(X509Certificate[] chain, String authType, SSLEngine engine)
+    throws CertificateException {
+    x509ExtendedTrustManager.checkServerTrusted(chain, authType, engine);
+    if (serverHostnameVerificationEnabled) {
+      try {
+        performHostVerification(InetAddress.getByName(engine.getPeerHost()), chain[0]);
+      } catch (UnknownHostException e) {
+        throw new CertificateException("Failed to verify host", e);
+      }
+    }
+  }
+
+  @Override
+  public void checkClientTrusted(X509Certificate[] chain, String authType)
+    throws CertificateException {
+    x509ExtendedTrustManager.checkClientTrusted(chain, authType);
+  }
+
+  @Override
+  public void checkServerTrusted(X509Certificate[] chain, String authType)
+    throws CertificateException {
+    x509ExtendedTrustManager.checkServerTrusted(chain, authType);
+  }
+
+  /**
+   * Compares peer's hostname with the one stored in the provided client certificate. Performs
+   * verification with the help of provided HostnameVerifier.
+   * @param inetAddress Peer's inet address.
+   * @param certificate Peer's certificate
+   * @throws CertificateException Thrown if the provided certificate doesn't match the peer
+   *                              hostname.
+   */
+  private void performHostVerification(InetAddress inetAddress, X509Certificate certificate)
+    throws CertificateException {
+    String hostAddress = "";
+    String hostName = "";
+    try {
+      hostAddress = inetAddress.getHostAddress();
+      hostnameVerifier.verify(hostAddress, certificate);
+    } catch (SSLException addressVerificationException) {
+      try {
```

Comment:
For a client, usually we need to use reverse dns to get its hostname, but for a server, usually we will use a hostname when connecting, so we should just check the host name first? But probably the algo here only adds an extra verification, as we will create the InetAddress with hostname so there is no reverse lookup when calling inetAddress.getHostName.

Reply:
Sorry @Apache9, I'm not entirely following your question/suggestion.

Are you suggesting that I change to use inetAddress.getHostName() here which uses reverse lookup? Or are you suggesting that I swap the order of the try/catch here so we try the reverse dns below first?

Since we are in the TrustManager layer, we only have what's available in this layer. So it's possible this method will be called in one of two ways:

- Two of the callers above (one for client, one for server), they pass the `socket.getInetAddress()`. This will probably be an ip address I guess.
- The other two callers above pass the parsed `engine.getPeerHost()` which according to the javadoc is the un-verified hostname.

The internals of SSL are hard to follow; I don't know in what circumstances each caller will be used. But both should work, because we have this 2 layer check (one direct check, one reverse lookup).

I'm happy to optimize this by changing the order if that's what you're asking. Please clarify if possible.

Reply:
I mean for verifying a server, usually we should check host name first, before checking the ip, as checking ip requires a dns lookup. But since we will finally connect to the server, a dns lookup is always needed, so this should be fine.

Reply:
But anyway, reverse dns lookup is a bit expensive under some scenario. From my experience, if there is no valid response, a dns lookup can hang the program for several seconds...

So it is possible to add a config to disable reverse dns lookup for client verification?